### PR TITLE
COMPUTE-413: dx toolkit support for dx-jobutil-get-identity-token

### PR DESCRIPTION
**Pull request by bayou-brogrammer**
JIRA = COMPUTE-413
---

Good portion of the changes are from running the generators. The 3 main files to focus on:

- src/python/dxpy/api.py
- src/python/test/test_dx_bash_helpers.py
- src/python/scripts/dx-jobutil-get-identity-token

### CODE REVIEW

**Review by saisreepamidi — APPROVED** [review]
At a high level it looks good to me, but I think its important to gets Kurts approval on this before you merge it, I lack domain knowledge around the CLI.

File: src/python/test/test_dx_bash_helpers.py [hunk 1]
```diff
@@ -269,41 +342,77 @@ def test_xattr_parameters(self):
...
 
             # Run the applet
             applet_args = ["-iseq1=A.txt", "-iseq2=B.txt", "-iref=A.txt", "-iref=B.txt"]
-            cmd_args = ['dx', 'run', '--yes', '--watch', applet_id]
+            cmd_args = ["dx", "run", "--yes", "--watch", applet_id]
             cmd_args.extend(applet_args)
             run(cmd_args, env=env)
 
-    @unittest.skipUnless(testutil.TEST_RUN_JOBS, 'skipping test that would run a job')
+    def test_job_identity_token(self):
```
Comment (wormsik) [inline src/python/test/test_dx_bash_helpers.py:355]: I would consider creating a new test suite for tokens...I don't feel that this is the right place...what do you think @kpjensen ?

File: src/python/scripts/dx-jobutil-get-identity-token [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+parser = argparse.ArgumentParser(
+    description="Creates a new job to run the named function with the specified input.  If successful, prints the ID of the new job."
+)
+parser.add_argument("--aud", help="Audience URI the JWT is intended for", required=True)
+parser.add_argument(
+    "--subject_claims",
+    action="append",
+    metavar="<subject_claims>",
+    help="Defines the subject claims to be validated by the cloud provider",
```
Comment (wormsik) [inline src/python/scripts/dx-jobutil-get-identity-token:30]: can you please describe also supported formats of providint subject claims in help? it seems that you can either use this arg multiple times or put multiple claims and separate using comma...

File: src/python/dxpy/toolkit_version.py [hunk 3]
```diff
@@ -1 +1 @@
-version = '0.370.0'
+version = '0.370.0-13-gb5b7342e6'
```
Comment (wormsik) [inline src/python/dxpy/toolkit_version.py:1]: can you please revert version? I think we no longer use automatic bump...am I right @kpjensen ?

File: src/python/test/file_load/job_identity_token/README.md
Comment (wormsik) [inline src/python/test/file_load/job_identity_token/README.md:1]: can you please justify why is this applet saved in `file_load` directory?

File: src/python/test/file_load/job_identity_token/dxapp.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  "runSpec": {
+    "file": "run.sh",
+    "interpreter": "bash",
+    "distribution": "Ubuntu",
+    "release": "20.04",
+    "version": "0"
+  },
+  "inputSpec": [
+    {"name": "audience",  "class": "string"},
+    {"name": "subject_claims",  "class": "array:string", "optional" : true}
```
Comment (wormsik) [inline src/python/test/file_load/job_identity_token/dxapp.json:13]: I would consider removing optional as if not provided, you'd get `subject_claims = ['']` which is not valid request

File: src/python/scripts/dx-jobutil-get-identity-token [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+# Parse out audience and subject_claims from args
+if args.aud is not None:
+    job_get_identity_token_input["audience"] = args.aud
+if args.subject_claims is not None:
+    # Iterate over subject_claims and flatten them into a single list
+    subject_claims = []
+    for subject_claim in args.subject_claims:
+        subject_claims.extend(subject_claim.split(","))
+    job_get_identity_token_input["subjectClaims"] = subject_claims
```
Comment (wormsik) [inline src/python/scripts/dx-jobutil-get-identity-token:44]: shouldn't be the request field name `subject_claims`?

File: src/python/test/test_dx_bash_helpers.py [hunk 6]
```diff
@@ -269,41 +342,77 @@ def test_xattr_parameters(self):
...
+        with temporary_project(
+            "TestDXBashHelpers.test_app1 temporary project"
+        ) as dxproj:
+            env = update_environ(DX_PROJECT_CONTEXT_ID=dxproj.get_id())
+
+            # Build the applet, patching in the bash helpers from the
+            # local checkout
+            applet_id = build_app_with_bash_helpers(
+                os.path.join(TEST_APPS, "job_identity_token"), dxproj.get_id()
+            )
```
Comment (wormsik) [inline src/python/test/test_dx_bash_helpers.py:366]: not sure if we have to use patched bash helpers here...perhaps just plain `dx build` would do the trick? (so it is not dependent on this directory)

**Review by wormsik — CHANGES_REQUESTED** [review]

File: src/python/scripts/dx-jobutil-get-identity-token (hunk 5 above)
Comment (bayou-brogrammer) [reply src/python/scripts/dx-jobutil-get-identity-token:44]: on the javascript side everything is `camelCase` but python is `snake case`. We decided as a team to have python keep python standards and the api adhere to js standards

File: src/python/test/test_dx_bash_helpers.py (hunk 6 above)
Comment (bayou-brogrammer) [reply src/python/test/test_dx_bash_helpers.py:366]: this is where I was told to place the test. Its an external bash script to dx toolkit so i think having a test in bash_helpers makes sense

File: src/python/test/test_dx_bash_helpers.py (hunk 1 above)
Comment (kpjensen) [reply src/python/test/test_dx_bash_helpers.py:357]: I think it's okay to leave here as similar python standalone test scripts for `dx-upload-all-outputs` exist here.

File: src/python/test/test_dx_bash_helpers.py (hunk 6 above)
Comment (kpjensen) [reply src/python/test/test_dx_bash_helpers.py]: All of the standalone scripts are referred to as "bash helpers", even though this one is pure python. This helper ensures the remote applet installs `dxpy` to be the same as the checkout running the tests.

File: src/python/test/jwt/README.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+This is a testing app, for the dx-upload/download-all python scripts.
```
Comment (kpjensen) [inline src/python/test/jwt/README.md:1]:
```suggestion
This is a testing app for the dx-jobutil-get-identity-token script. 
```

**Review by wormsik — APPROVED** [review]

**Comment by wormsik** [comment]
@bayou-brogrammer Thanks for addressing my comments! I have nothing left on my side and just let final approval on @kpjensen .

**Pull request merged**